project_euler/problem_345/sol1.py

```diff
@@ -0,0 +1,176 @@
+    [7, 53, 183, 439, 863],
+    [497, 383, 563, 79, 973],
+    [287, 63, 343, 169, 583],
+    [627, 343, 773, 959, 943],
+    [767, 473, 103, 699, 303],
+]
+
+MATRIX_2: list = [
+    [7, 53, 183, 439, 863, 497, 383, 563, 79, 973, 287, 63, 343, 169, 583],
+    [627, 343, 773, 959, 943, 767, 473, 103, 699, 303, 957, 703, 583, 639, 913],
+    [447, 283, 463, 29, 23, 487, 463, 993, 119, 883, 327, 493, 423, 159, 743],
+    [217, 623, 3, 399, 853, 407, 103, 983, 89, 463, 290, 516, 212, 462, 350],
+    [960, 376, 682, 962, 300, 780, 486, 502, 912, 800, 250, 346, 172, 812, 350],
+    [870, 456, 192, 162, 593, 473, 915, 45, 989, 873, 823, 965, 425, 329, 803],
+    [973, 965, 905, 919, 133, 673, 665, 235, 509, 613, 673, 815, 165, 992, 326],
+    [322, 148, 972, 962, 286, 255, 941, 541, 265, 323, 925, 281, 601, 95, 973],
+    [445, 721, 11, 525, 473, 65, 511, 164, 138, 672, 18, 428, 154, 448, 848],
+    [414, 456, 310, 312, 798, 104, 566, 520, 302, 248, 694, 976, 430, 392, 198],
+    [184, 829, 373, 181, 631, 101, 969, 613, 840, 740, 778, 458, 284, 760, 390],
+    [821, 461, 843, 513, 17, 901, 711, 993, 293, 157, 274, 94, 192, 156, 574],
+    [34, 124, 4, 878, 450, 476, 712, 914, 838, 669, 875, 299, 823, 329, 699],
+    [815, 559, 813, 459, 522, 788, 168, 586, 966, 232, 308, 833, 251, 631, 107],
+    [813, 883, 451, 509, 615, 77, 281, 613, 459, 205, 380, 274, 302, 35, 805],
+]
+
+
+def solution(
+    matrix: list = MATRIX_2,
+    enhanced_matrix: list = None,
+    i: int = 0,
+    booked_list: list = None,
+    value: int = 0,
+    max_value: float = float("-inf"),
+):
+    """
+    Returns a maximum possible sum of matrix values
+    with no value shares a same row or column.
+
+    >>> solution(MATRIX_1)
+    3315
+
+    Arguments:
+    matrix              --  Original matrix from input
+    enhanced_matrix     --  Original matrix but each row is sorted and each cell
+                            in tuple form and each value is paired
+                            with its column index
+    i                   --  i-th row
+    booked_list         --  List of j's flag
+    value               --  Accumulative value from parent to current node
+    max_value           --  Maximum value is stored right now by the tree
+    """
+
+    # Get length of a row and length of a column, respectively denoted by m and n
+    m, n = len(matrix), len(matrix[0])
+
+    # Create list of j's flag at the first time
+    if booked_list is None:
+        booked_list = [False for _ in range(m)]
+
+    # Create enhanced matrix at the first time
+    if enhanced_matrix is None:
+        # Pair value with its column index
+        enhanced_matrix = [[(row[j], j) for j in range(len(row))] for row in matrix]
+
+        # Sort each row by considering the value as comparison
+        for k in range(len(enhanced_matrix)):
+            enhanced_matrix[k] = sorted(
+                enhanced_matrix[k],
+                key=cmp_to_key(lambda a, b: 1 if a[0] < b[0] else -1),
```

Comment:
Can you explain the use case of `cmp_to_key` in here? According to the Contributing guidelines we avoid the use of `lambda` so I am just trying to understand what's going on in here.

Reply:
Just to make it clear, so each row in enchance_matrix is a list of a tuple and each tuple contains two elements. Those codes tell to sort each row in ascending order by considering the first element as a comparison. I'm not sure what cmp_to_key is doing but in order to sort by looking at the first element only, I need to use lambda and ```cmp_to_key``` to make it work in the ```sorted``` function. I don't know about another way to replace this behavior, but if you have one let me know so I can change it and will follow the guideline. Thank you!

Reply:
I just realized that NumPy can do the work, yea I think I will use NumPy for the array

Reply:
Fixed: I will use NumPy for each row only, not the whole matrix